nexus/db-model/src/affinity.rs

```diff
@@ -0,0 +1,85 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/5.0/.
+
+// Copyright 2024 Oxide Computer Company
+
+//! Database representation of affinity and anti-affinity groups
+
+use super::impl_enum_type;
+use crate::schema::affinity_group;
+use crate::schema::affinity_group_instance_membership;
+use crate::schema::anti_affinity_group;
+use crate::schema::anti_affinity_group_instance_membership;
+use crate::typed_uuid::DbTypedUuid;
+use chrono::{DateTime, Utc};
+use omicron_uuid_kinds::AffinityGroupKind;
+use omicron_uuid_kinds::AntiAffinityGroupKind;
+use omicron_uuid_kinds::InstanceKind;
+
+impl_enum_type!(
+    #[derive(SqlType, Debug)]
+    #[diesel(postgres_type(name = "affinity_policy", schema = "public"))]
+    pub struct AffinityPolicyEnum;
+
+    #[derive(Clone, Copy, Debug, AsExpression, FromSqlRow, PartialEq)]
+    #[diesel(sql_type = AffinityPolicyEnum)]
+    pub enum AffinityPolicy;
+
+    // Enum values
+    Fail => b"fail"
+    Allow => b"allow"
+);
+
+impl_enum_type!(
+    #[derive(SqlType, Debug)]
+    #[diesel(postgres_type(name = "affinity_distance", schema = "public"))]
+    pub struct AffinityDistanceEnum;
+
+    #[derive(Clone, Copy, Debug, AsExpression, FromSqlRow, PartialEq)]
+    #[diesel(sql_type = AffinityDistanceEnum)]
+    pub enum AffinityDistance;
+
+    // Enum values
+    Sled => b"sled"
+);
+
+#[derive(Queryable, Insertable, Clone, Debug, Selectable)]
+#[diesel(table_name = affinity_group)]
+pub struct AffinityGroup {
+    pub id: DbTypedUuid<AffinityGroupKind>,
+    pub name: String,
+    pub description: String,
+    pub time_created: DateTime<Utc>,
+    pub time_modified: DateTime<Utc>,
+    pub time_deleted: Option<DateTime<Utc>>,
+    pub policy: AffinityPolicy,
+    pub distance: AffinityDistance,
+}
+
+#[derive(Queryable, Insertable, Clone, Debug, Selectable)]
+#[diesel(table_name = anti_affinity_group)]
+pub struct AntiAffinityGroup {
+    pub id: DbTypedUuid<AntiAffinityGroupKind>,
+    pub name: String,
+    pub description: String,
+    pub time_created: DateTime<Utc>,
+    pub time_modified: DateTime<Utc>,
+    pub time_deleted: Option<DateTime<Utc>>,
+    pub policy: AffinityPolicy,
+    pub distance: AffinityDistance,
+}
+
+#[derive(Queryable, Insertable, Clone, Debug, Selectable)]
+#[diesel(table_name = affinity_group_instance_membership)]
+pub struct AffinityGroupInstanceMembership {
+    pub group_id: DbTypedUuid<AffinityGroupKind>,
+    pub instance_id: DbTypedUuid<InstanceKind>,
+}
+
+#[derive(Queryable, Insertable, Clone, Debug, Selectable)]
+#[diesel(table_name = anti_affinity_group_instance_membership)]
+pub struct AntiAffinityGroupInstanceMembership {
+    pub group_id: DbTypedUuid<AntiAffinityGroupKind>,
+    pub instance_id: DbTypedUuid<InstanceKind>,
+}
```

Comment:
I note that these lack created/deleted timestamps, implying that:
1. we intend to hard-delete rather than soft-delete them, and,
2. we don't presently record when instances were added to affinity/anti-affinity groups, so we can't present that in UIs in the future. 

I'm not sure if either of these matter to us, but I figured I'd comment on it.

Reply:
When I wrote this I was kinda planning on just using hard deletion for membership. I *could* definitely add the "time_modified" and "time_deleted" columns, and use soft-deletion here too, but as usual, we'll need to be more cautious with our indexing.

("Why hard delete" -> this was kinda arbitrary, my decision here isn't super strong, but I'm more familiar with us using soft deletion for user-facing objects that have full CRUD APIs, and hard-deletion for more internal-facing stuff, to avoid the cost of garbage collecting later, which we haven't really done at all)

Reply:
That's totally fair!

The reason I brought up deletion was because if we start out with a schema that uses hard-deletion, and decide to switch to soft-deletion later in order to do something like display affinity group histories in the UI, we can't get back records from before that change, since...they've been deleted. On the other hand, if we started with soft-deletion, we could switch to hard-deletion and blow away any soft-deleted records if we decide to not use them in that way.

On the other hand, maybe the problem of displaying historical affinity group changes is better solved by other things, like audit logging! I dunno.